src/net/http_response.c

```diff
@@ -0,0 +1,366 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <postgres.h>
+#include <lib/stringinfo.h>
+#include <utils/memutils.h>
+
+#include "http.h"
+
+#define CARRIAGE_RETURN	'\r'
+#define NEW_LINE		'\n'
+#define SEP_CHAR		':'
+#define HTTP_VERSION_BUFFER_SIZE 128
+
+extern HttpHeader *http_header_create(const char *name, size_t name_len, const char *value, size_t value_len, HttpHeader *next);
+
+typedef enum HttpParseState
+{
+	HTTP_STATE_STATUS,
+	HTTP_STATE_INTERM,			/* received a single \r */
+	HTTP_STATE_HEADER_NAME,		/* received \r\n */
+	HTTP_STATE_HEADER_VALUE,
+	HTTP_STATE_ALMOST_DONE,
+	HTTP_STATE_BODY,
+	HTTP_STATE_ERROR,
+	HTTP_STATE_DONE,
+} HttpParseState;
+
+typedef struct HttpResponseState
+{
+	MemoryContext context;
+	char		version[HTTP_VERSION_BUFFER_SIZE];
+	char		raw_buffer[MAX_RAW_BUFFER_SIZE];
+	/* The next read should copy data into the buffer starting here */
+	off_t		offset;
+	off_t		parse_offset;
+	size_t		cur_header_name_len;
+	size_t		cur_header_value_len;
+	char	   *cur_header_name;
+	char	   *cur_header_value;
+	HttpHeader *headers;
+	int			status_code;
+	size_t		content_length;
+	char	   *body_start;
+	HttpParseState state;
+} HttpResponseState;
+
+void
+http_response_state_init(HttpResponseState *state)
+{
+	state->status_code = -1;
+	state->state = HTTP_STATE_STATUS;
+}
+
+HttpResponseState *
+http_response_state_create()
+{
+	MemoryContext context = AllocSetContextCreate(CurrentMemoryContext,
+												  "Http Response",
+												  ALLOCSET_DEFAULT_SIZES);
+	MemoryContext old = MemoryContextSwitchTo(context);
+	HttpResponseState *ret = palloc(sizeof(HttpResponseState));
+
+	memset(ret, 0, sizeof(*ret));
+
+	ret->context = context;
+	http_response_state_init(ret);
+	MemoryContextSwitchTo(old);
+	return ret;
+}
+
+void
+http_response_state_destroy(HttpResponseState *state)
+{
+	MemoryContextDelete(state->context);
+}
+
+bool
+http_response_state_valid_status(HttpResponseState *state)
+{
+	/* If the status code hasn't been parsed yet, return */
+	if (state->status_code == -1)
+		return true;
+	/* If it's a bad status code, then bad! */
+	if (state->status_code / 100 == 2)
+		return true;
+	return false;
+}
+
+bool
+http_response_state_is_done(HttpResponseState *state)
+{
+	return (state->state == HTTP_STATE_DONE);
+}
+
+size_t
+http_response_state_buffer_remaining(HttpResponseState *state)
+{
+	Assert(state->offset <= MAX_RAW_BUFFER_SIZE);
```

Comment:
This checks for bad state (== bug) so should not be an error in my mind. The contract here should be to return the remaining buffer state or zero if none. It should never return error.

Reply:
The only place where `state->offset` is updated is in `http_response_state_parse()`:

```
state->offset += bytes;

	if (state->offset > MAX_RAW_BUFFER_SIZE)
		state->offset = MAX_RAW_BUFFER_SIZE;
```

I don't see how the thing you mention is possible, and we do `Assert()` it. But if you insist that we need to have errors for impossible cases, then sure I can add that.

Reply:
Logging added.

FWIW, I believe the `Assert()` fully provides both the needed in-code "documentation" and test case for protecting against future intern screw ups. Otherwise, if doubts remain on bugs or lack of coverage for future regressions, we should, IMHO, focus on improving tests instead of riddling the code with `elog` statements. I guess we just need to determine when to use `elog` vs `Assert()`, because otherwise there are numerous `Assert()`ed and impossible cases in the code you could make the same future-intern-screwup argument around.